tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/XUnit/XUnitTestAssemblyRunnerRunAsyncIntegration.cs

```diff
@@ -0,0 +1,114 @@
+
+/// <summary>
+/// Xunit.Sdk.TestAssemblyRunner`1.RunTestCollectionAsync calltarget instrumentation
+/// </summary>
+[InstrumentMethod(
+    AssemblyNames = new[] { "xunit.execution.dotnet", "xunit.execution.desktop" },
+    TypeName = "Xunit.Sdk.TestAssemblyRunner`1",
+    MethodName = "RunAsync",
+    ReturnTypeName = "System.Threading.Tasks.Task`1<Xunit.Sdk.RunSummary>",
+    ParameterTypeNames = new string[0],
+    MinimumVersion = "2.2.0",
+    MaximumVersion = "2.*.*",
+    IntegrationName = XUnitIntegration.IntegrationName)]
+[Browsable(false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+public static class XUnitTestAssemblyRunnerRunAsyncIntegration
+{
+    /// <summary>
+    /// OnMethodBegin callback
+    /// </summary>
+    /// <typeparam name="TTarget">Type of the target</typeparam>
+    /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+    /// <returns>Calltarget state value</returns>
+    internal static CallTargetState OnMethodBegin<TTarget>(TTarget instance)
+    {
+        if (!XUnitIntegration.IsEnabled || instance is null)
+        {
+            return CallTargetState.GetDefault();
+        }
+
+        var assemblyRunnerInstance = instance.DuckCast<TestAssemblyRunnerStruct>();
+        if (assemblyRunnerInstance.TestAssembly.Assembly.Name is { } assemblyName)
+        {
+            var testBundleString = new AssemblyName(assemblyName).Name ?? string.Empty;
+
+            // Extract the version of the framework from the TestClassRunner base class
+            var frameworkType = instance.GetType();
+            while (frameworkType.IsAbstract == false)
+            {
+                if (frameworkType.BaseType is { } baseType)
+                {
+                    frameworkType = baseType;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            CIVisibility.WaitForSkippableTaskToFinish();
+            return new CallTargetState(null, TestModule.Create(testBundleString, "xUnit", frameworkType.Assembly.GetName().Version?.ToString()));
+        }
+
+        return CallTargetState.GetDefault();
+    }
+
+    /// <summary>
+    /// OnMethodEnd callback
+    /// </summary>
+    /// <typeparam name="TTarget">Type of the target</typeparam>
+    /// <typeparam name="TResult">TestResult type</typeparam>
+    /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+    /// <param name="returnValue">Original method return value</param>
+    /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+    /// <param name="state">Calltarget state value</param>
+    /// <returns>Return value of the method</returns>
+    internal static CallTargetReturn<TResult> OnMethodEnd<TTarget, TResult>(TTarget instance, TResult returnValue, Exception exception, in CallTargetState state)
+    {
+        if (state.State == TestModule.Current)
```

Comment:
Good find, I wonder if there's a way we can improve this so it could be done automatically or we could have an automated test for it

Reply:
It's hard to do this automatically, CallTarget instrumentation already does this for the Scope. 

https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/EndMethodHandler%601.cs#L74-L85

The problem is that the Scope is tied to a Span. In this case we have a different `AsyncLocal` in other type. But we want to have the same behavior as the Scope so we don't leak the change if the Task is not awaited. Unfortunately doing a generic way to handle this is outside the scope of this PR.